partiql-types/src/main/kotlin/org/partiql/value/PartiQLValue.kt

```diff
@@ -0,0 +1,200 @@
+        get() = value.toDouble()
+
+    val double: Float
+        get() = value.toFloat()
+
+    override fun toString(): String = value.toString()
+}
+
+abstract class Int8Value : NumericValue<Byte>() {
+
+    override val type: PartiQLType = PartiQLType.INT8
+}
+
+abstract class Int16Value : NumericValue<Short>() {
+
+    override val type: PartiQLType = PartiQLType.INT16
+}
+
+abstract class Int32Value : NumericValue<Int>() {
+
+    override val type: PartiQLType = PartiQLType.INT32
+}
+
+abstract class Int64Value : NumericValue<Long>() {
+
+    override val type: PartiQLType = PartiQLType.INT64
+}
+
+abstract class IntValue : NumericValue<BigInteger>() {
+
+    override val type: PartiQLType = PartiQLType.INT
+}
+
+abstract class DecimalValue : NumericValue<BigDecimal>() {
+
+    override val type: PartiQLType = PartiQLType.DECIMAL
+}
+
+abstract class Float32Value : ScalarValue<Float> {
+
+    override val type: PartiQLType = PartiQLType.FLOAT32
+}
+
+abstract class Float64Value : ScalarValue<Double> {
+
+    override val type: PartiQLType = PartiQLType.FLOAT64
+}
+
+sealed class TextValue<T> : ScalarValue<T> {
+
+    abstract val string: String
+
+}
+
+abstract class CharValue : TextValue<Char>() {
+
+    override val type: PartiQLType = PartiQLType.CHAR
+
+    override val string: String
+        get() = type.toString()
+}
+
+abstract class StringValue : TextValue<String>() {
+
+    override val type: PartiQLType = PartiQLType.STRING
+
+    override val string: String
+        get() = value
+}
+
+abstract class BitValue : ScalarValue<Boolean> {
+
+    override val type: PartiQLType = PartiQLType.BIT
+}
+
+abstract class BinaryValue : ScalarValue<BitSet> {
+
+    override val type: PartiQLType = PartiQLType.BINARY
+}
+
+abstract class ByteValue : ScalarValue<Byte> {
+
+    override val type: PartiQLType = PartiQLType.BYTE
+}
+
+abstract class BlobValue : ScalarValue<ByteArray> {
+
+    override val type: PartiQLType = PartiQLType.BLOB
+}
+
+abstract class DateValue : ScalarValue<Date> {
+
+    override val type: PartiQLType = PartiQLType.DATE
+}
+
+abstract class TimeValue : ScalarValue<Long> {
+
+    override val type: PartiQLType = PartiQLType.TIME
+}
+
+abstract class TimestampValue : ScalarValue<Instant> {
+
+    override val type: PartiQLType = PartiQLType.TIMESTAMP
+}
+
+abstract class IntervalValue : ScalarValue<Long> {
+
+    override val type: PartiQLType = PartiQLType.INTERVAL
+}
+
+abstract class BagValue<T : PartiQLValue> : CollectionValue<T> {
+
+    override val type: PartiQLType = PartiQLType.BAG
+}
+
+abstract class ArrayValue<T : PartiQLValue> : CollectionValue<T> {
+
+    override val type: PartiQLType = PartiQLType.ARRAY
+}
+
+abstract class TupleValue<T : PartiQLValue> : PartiQLValue, Collection<Pair<String, T>> {
+
+    abstract val fields: List<Pair<String, T>>
+
+    override val type: PartiQLType = PartiQLType.TUPLE
+}
+
+/**
+ * Any view over a PartiQLValue
+ */
+abstract class AnyValue : PartiQLValue {
```

Comment:
I see the resemblance to AnyElement -- and I'm curious what benefits Ion has gained from using this.